crates/sui/tests/framework_upgrades/mock_sui_systems/deep_upgrade/sources/validator_wrapper.move

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module sui_system::validator_wrapper {
+    use sui::versioned::Versioned;
+    use sui::versioned;
+    use sui::tx_context::TxContext;
+    use sui_system::validator::{Validator, ValidatorV3};
+    use sui_system::validator;
+
+    friend sui_system::sui_system;
+    friend sui_system::sui_system_state_inner;
+
+    const VALIDATOR_VERSION_V1: u64 = 18446744073709551605;  // u64::MAX - 10
+    const VALIDATOR_VERSION_V3: u64 = 18446744073709551607;  // u64::MAX - 8
+
+    const EInvalidVersion: u64 = 0;
+
+    struct ValidatorWrapper has store {
+        inner: Versioned
+    }
+
+    // Validator corresponds to version 1.
+    public(friend) fun create_v1(validator: Validator, ctx: &mut TxContext): ValidatorWrapper {
+        ValidatorWrapper {
+            inner: versioned::create(VALIDATOR_VERSION_V1, validator, ctx)
+        }
+    }
+
+    /// This function should always return the latest supported version.
+    /// If the inner version is old, we upgrade it lazily in-place.
+    public(friend) fun load_validator_maybe_upgrade(self: &mut ValidatorWrapper): &mut ValidatorV3 {
+        upgrade_to_latest(self);
+        versioned::load_value_mut(&mut self.inner)
+    }
+
+    /// Destroy the wrapper and retrieve the inner validator object.
+    public(friend) fun destroy(self: ValidatorWrapper): ValidatorV3 {
+        upgrade_to_latest(&mut self);
+        let ValidatorWrapper { inner } = self;
+        versioned::destroy(inner)
+    }
+
+    fun upgrade_to_latest(self: &mut ValidatorWrapper) {
```

Comment:
Since this is lazy and a validator wrapper may not be touched for quite a few epochs where multiple upgrades may have happened, this means that we will have to keep all the migration functions `v1_to_v2`, `v2_to_v3`, `v3_to_v4`, etc, and possibly perform more than one upgrades in this function. Is that right?

Reply:
Correct.